gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/common/impl/InMemoryTopology.java

```diff
@@ -0,0 +1,346 @@
+        "Root can not import other nodes, otherwise circular dependency will happen");
+
+    while (!currentLevel.isEmpty()) {
+      Collection<ConfigKeyPath> nextLevel = new ArrayList<ConfigKeyPath>();
+      for (ConfigKeyPath configKeyPath : currentLevel) {
+        // calls to retrieve cache / set cache if not present
+        List<ConfigKeyPath> ownImports = this.getOwnImports(configKeyPath);
+        for (ConfigKeyPath p : ownImports) {
+          this.addToCollectionMapForSingleValue(this.ownImportedByMap, p, configKeyPath);
+        }
+
+        // calls to retrieve cache / set cache if not present
+        Collection<ConfigKeyPath> tmp = this.getChildren(configKeyPath);
+        nextLevel.addAll(tmp);
+      }
+      currentLevel = nextLevel;
+    }
+
+    // traversal the ownImports map to get the recursive imports map and set recursive imported by map
+    for (ConfigKeyPath configKey : this.ownImportMap.keySet()) {
+      List<ConfigKeyPath> recursiveImports = null;
+      // result may in the cache already 
+      if (this.recursiveImportMap.containsKey(configKey)) {
+        recursiveImports = this.recursiveImportMap.get(configKey);
+      } else {
+        recursiveImports = this.buildImportsRecursiveFromCache(configKey);
+        this.addToListMapForListValue(recursiveImportMap, configKey, recursiveImports);
+      }
+
+      if (recursiveImports != null) {
+        for (ConfigKeyPath p : recursiveImports) {
+          this.addToCollectionMapForSingleValue(this.recursiveImportedByMap, p, configKey);
+        }
+      }
+    }
+  }
+
+  private List<ConfigKeyPath> buildImportsRecursiveFromCache(ConfigKeyPath configKey) {
+    return buildImportsRecursiveFromCacheHelper(configKey, configKey, new ArrayList<ConfigKeyPath>());
+  }
+
+  private List<ConfigKeyPath> buildImportsRecursiveFromCacheHelper(ConfigKeyPath initialConfigKey,
+      ConfigKeyPath currentConfigKey, List<ConfigKeyPath> previous) {
+
+    for (ConfigKeyPath p : previous) {
+      if (currentConfigKey != null && currentConfigKey.equals(p)) {
+        previous.add(p);
+        throw new CircularDependencyException(getCircularDependencyChain(initialConfigKey, previous, currentConfigKey));
+      }
+    }
+
+    // root can not include anything, otherwise will have circular dependency
+    if (currentConfigKey.isRootPath()) {
+      return Collections.emptyList();
+    }
+
+    List<ConfigKeyPath> result = new ArrayList<>();
+
+    // Own imports map should be filled in already
+    List<ConfigKeyPath> rawList = this.getOwnImportsFromCache(currentConfigKey);
+
+    // without this wrapper, will cause UnsupportedOperationException in adding parent node
+    List<ConfigKeyPath> imported = new ArrayList<ConfigKeyPath>();
+    imported.addAll(rawList);
+
+    // implicit import parent with the lowest priority
+    if (!currentConfigKey.isRootPath()) {
+      imported.add(currentConfigKey.getParent());
+    }
+
+    for (ConfigKeyPath u : imported) {
```

Comment:
I think you can save the allocation for imported by moving the loop body in a method, say addRecursiveImportsToResult();

Then, your code might look like:

for (ConfigKeyPath u: this.getOwnImportsFromCache(currentConfigKey)) {
  addRecursiveImportsToResult(u, result, current);
}
addRecursiveImportsToResult(currentConfigKey.getParent(), result, current);